language-plutus-core/src/Language/PlutusCore/Type.hs

```diff
@@ -91,29 +93,57 @@ instance Corecursive (Type tyname a) where
     embed (TyLamF l tn k ty)    = TyLam l tn k ty
     embed (TyAppF l ty ty')     = TyApp l ty ty'
 
--- | Substitute @a@ for @b@ in @c@.
-tyNameSubstitute :: Eq (tyname a)
-                 => tyname a -- ^ @a@
-                 -> tyname a -- ^ @b@
-                 -> Type tyname a -- ^ @c@
-                 -> Type tyname a
-tyNameSubstitute tn tn' = cata a where
-    a (TyVarF x tn'') | tn == tn'' = TyVar x tn'
-    a x               = embed x
-
-instance (Eq (tyname a), Eq a) => Eq (Type tyname a) where
-    (==) (TyVar a1 tn) (TyVar a2 tn')                   = a1 == a2 && tn == tn'
-    (==) (TyFun a1 ty ty') (TyFun a2 ty'' ty''')        = a1 == a2 && ty == ty'' && ty' == ty'''
-    (==) (TyFix a1 tn ty) (TyFix a2 tn' ty')            =
-        a1 == a2 && tyNameSubstitute tn' tn ty' == ty
-    (==) (TyForall a1 tn k ty) (TyForall a2 tn' k' ty') =
-        a1 == a2 && tyNameSubstitute tn' tn ty' == ty && k == k'
-    (==) (TyBuiltin a1 b) (TyBuiltin a2 b')             = a1 == a2 && b == b'
-    (==) (TyInt a1 n) (TyInt a2 n')                     = a1 == a2 && n == n'
-    (==) (TyLam a1 tn k ty) (TyLam a2 tn' k' ty')       =
-        a1 == a2 && tyNameSubstitute tn' tn ty' == ty && k == k'
-    (==) (TyApp a1 ty ty') (TyApp a2 ty'' ty''')        = a1 == a2 && ty == ty'' && ty' == ty'''
-    (==) _ _                                            = False
+-- this type is used for replacing type names in
+-- the Eq instance
+type EqState tyname a = M.Map (tyname a) (tyname a)
+
+rebind :: (Ord (tyname a), MonadState (EqState tyname a) m)
+       => tyname a
+       -> tyname a
+       -> m ()
+rebind = modify .* M.insert
+
+rebindAndEq :: (Eq a, Ord (tyname a), MonadState (EqState tyname a) m)
+            => Type tyname a
+            -> Type tyname a
+            -> tyname a
+            -> tyname a
+            -> m Bool
+rebindAndEq ty ty' tn tn' =
+    rebind tn' tn *>
+    eqTypeM ty ty'
```

Comment:
Hmm... I feel like globally rebinding variables via the `State` monad for a local equality check should lead to bugs. I'll think more on this, please do not merge yet.

Reply:
I think this is a good idea.

Reply:
For the time being I'm using `remove` to drop a variable once it goes out of scope. 

I think explicit passing may be more efficient but this will do.

Reply:
> I think the simplest thing here is just not to use a monad. And instead pass environments explicitly as an argument.

I couldn't be more unclear. I mean, right now we use `State` to emulate `Reader`. You add a name to the current state and immediately delete it after a local equality check is performed. This is the behavior of `Reader` essentially. So, can't we simply use `Reader`? Or just a plain argument.

This also applies to the type checker as well, I believe (which makes me thinking about our whole way of dealing with environments, uniques and such).

Reply:
Yes, I have a ticket to use the argument-passing approach in the type-checker as well; this seems sensible here.